### PR TITLE
tree-wide: declare iterator of LIST_FOREACH() in the loop

### PR DESCRIPTION
**Pull request by yuwata**
Prompted by https://lwn.net/Articles/887097/

### CODE REVIEW

File: src/basic/list.h [hunk 1]
```diff
@@ -137,32 +137,32 @@
         (!(item)->name##_prev && !(item)->name##_next)
 
 #define LIST_FOREACH(name,i,head)                                       \
-        for ((i) = (head); (i); (i) = (i)->name##_next)
+        for (typeof(head) i = (head); i; i = i->name##_next)
```
Comment (bluca) [inline src/basic/list.h]: Can we use the `UNIQ_T` stuff here?

Comment (yuwata) [reply src/basic/list.h]: No, the iterator must be usable in the loop. Hence, the provided name must be used.

**Comment by yuwata** [comment]
There exists a problem in the current version. The following does not work.
```
typedef struct Foo Foo;
struct Foo {
        LIST_FIELDS(Foo, foo);
};

typedef struct Obj {
        LIST_HEAD(Foo, head);
} Obj;

void func(const Obj *obj) {
        LIST_FOREACH(foo, i, obj->head) {
                ...;
        }
}
```
But `LIST_FOREACH(foo, i, (Foo*) obj->head)` works.

Because, `typeof(obj->head)` is `Foo * const` instead of `const Foo *`...

File: src/basic/list.h (hunk 1 above)
Comment (bluca) [reply src/basic/list.h]: I see - but then nesting won't work?

Comment (yuwata) [reply src/basic/list.h]: Why? There is no variable implicitly declared. We can safely use the macro arbitrary level of nesting:
```
LIST_FOREACH(name1, i, head1)
    LIST_FOREACH(name2, j, head2)
        ...;
```

Comment (bluca) [reply src/basic/list.h]: ah yes, I misread it, sorry

**Comment by yuwata** [comment]
@poettering, @keszybz and @bluca WDYT about the basic of this PR? If you like this, then I will apply similar changes to other loop macros.

**Comment by bluca** [comment]
> @poettering, @keszybz and @bluca WDYT about the basic of this PR? If you like this, then I will apply similar changes to other loop macros.

I like it! Maybe explicitly document in list.h that the iterator is declared for you by the macro, since that could be somewhat unexpected, but otherwise looks like a good usability improvement to me

**Comment by yuwata** [comment]
All green now. PTAL.

File: src/libsystemd-network/sd-radv.c [hunk 2]
```diff
@@ -659,15 +654,18 @@ int sd_radv_add_prefix(sd_radv *ra, sd_radv_prefix *p) {
         return 0;
 }
 
-sd_radv_prefix *sd_radv_remove_prefix(sd_radv *ra,
-                                               const struct in6_addr *prefix,
-                                               unsigned char prefixlen) {
-        sd_radv_prefix *cur, *next;
+void sd_radv_remove_prefix(
```
Comment (bluca) [inline src/libsystemd-network/sd-radv.c]: Could you please split this function refactor in a separate commit? Lumped together with the tree-wide iterator change it's very easy to miss

File: src/resolve/resolved-dns-zone.c [hunk 3]
```diff
@@ -78,12 +78,13 @@ void dns_zone_flush(DnsZone *z) {
 }
 
 DnsZoneItem* dns_zone_get(DnsZone *z, DnsResourceRecord *rr) {
-        DnsZoneItem *i;
+        DnsZoneItem *first;
 
         assert(z);
         assert(rr);
 
-        LIST_FOREACH(by_key, i, hashmap_get(z->by_key, rr->key))
```
Comment (bluca) [inline src/resolve/resolved-dns-zone.c:86]: the fact that function calls cannot be inlined anymore also needs to be documented in the list header I think

**Review by bluca — APPROVED** [review]
Overall I like the change - just some minor nits. Not being able to inline function calls in the macro invocation anymore might be a the major gotcha, but might be worth it in the end.

File: src/resolve/resolved-dns-zone.c (hunk 3 above)
Comment (yuwata) [reply src/resolve/resolved-dns-zone.c:86]: Or, in this case, you can inline them like the following:
```
LIST_FOREACH(by_key, i, (DnsZoneItem*) hashmap_get(...))
```

Comment (bluca) [reply src/resolve/resolved-dns-zone.c:86]: If it's still possible, maybe do that to keep the diff to a minimum? There's other instances as well when a temporary new variable was added to avoid the inline function call

Comment (yuwata) [reply src/resolve/resolved-dns-zone.c:86]: Sure.

File: src/libsystemd-network/sd-radv.c (hunk 2 above)
Comment (yuwata) [reply src/libsystemd-network/sd-radv.c]: Done.

File: src/partition/repart.c [hunk 4]
```diff
@@ -4797,9 +4772,8 @@ static int resize_backing_fd(
 
 static int determine_auto_size(Context *c) {
         uint64_t sum;
-        Partition *p;
 
-        assert_se(c);
+        assert(c);
```
Comment (medhefgo) [inline src/partition/repart.c:4773]: `assert_se -> assert` intentional?

**Review by medhefgo — APPROVED** [review]
+1 From me just because of the lines deleted count alone :smile_cat:.

File: src/partition/repart.c (hunk 4 above)
Comment (yuwata) [reply src/partition/repart.c:4773]: Yes, this is not a test code. Maybe, I should split the commit for the change.

**Comment by poettering** [comment]
Note that we don't have the same underlying problem as the kernel people with this, as we don#t do this containerof() based logic they usually do with embedding. i.e. our stuff uses the containing type as primary type we use, and does not embedd a structure from which one would have to return to the containing type somehow.

That said, it looks worthwhile from a first glance, given how it makes the code shorter.

File: src/basic/list.h [hunk 5]
```diff
@@ -139,14 +139,17 @@
 /* The type of the iterator 'i' is automatically determined by the type of 'head', and declared in the
  * loop. Hence, do not declare the same variable in the outer scope. Sometimes, we set 'head' through
  * hashmap_get(). In that case, you need to explicitly cast the result. */
+#define LIST_FOREACH_WITH_NEXT(name,i,n,head)                           \
```
Comment (keszybz) [inline src/basic/list.h:142]: Shouldn't this be `_LIST_FOREACH_WITH_NEXT()` to match `_LIST_FOREACH_WITH_PREV()` ?

**Review by keszybz — COMMENTED** [review]
Looks great! Thank you for doing this.

File: src/basic/path-util.c [hunk 6]
```diff
@@ -202,9 +202,9 @@ int path_make_relative(const char *from, const char *to, char **ret) {
 }
 
 char* path_startswith_strv(const char *p, char **set) {
-        char **s, *t;
-
         STRV_FOREACH(s, set) {
+                char *t;
```
Comment (medhefgo) [inline src/basic/path-util.c:206]: Doesn't really matter, but as I understand it, the iterator is now safe to modify, so this t could be done away with, no?

File: src/basic/path-util.c [hunk 7]
```diff
@@ -1421,10 +1416,9 @@ bool path_strv_contains(char **l, const char *path) {
 }
 
 bool prefixed_path_strv_contains(char **l, const char *path) {
-        char **i, *j;
-
         STRV_FOREACH(i, l) {
-                j = *i;
+                char *j = *i;
```
Comment (medhefgo) [inline src/basic/path-util.c]: Same here.

File: src/basic/strv.h [hunk 8]
```diff
@@ -122,19 +122,28 @@ static inline int strv_from_nulstr(char ***a, const char *nulstr) {
...
+
 #define STRV_FOREACH(s, l)                      \
-        for ((s) = (l); (s) && *(s); (s)++)
+        _STRV_FOREACH(s, l, UNIQ_T(i, UNIQ))
+
+#define _STRV_FOREACH_BACKWARDS(s, l, h, i)                             \
+        for (typeof(*(l)) *s, *h = (l),                                 \
+                     *i = h ? h + strv_length(h) - 1U : NULL;           \
+             h && (s = i) >= h;                                         \
+             i--)
```
Comment (medhefgo) [inline src/basic/strv.h:137]: Not a total expert, but isn't creating a pointer that points (one element) before a malloc technically UB? (I know the code before this does this already.)

File: src/basic/path-util.c (hunk 6 above)
Comment (poettering) [reply src/basic/path-util.c:206]: the function doesn't modify the iterator though?

File: src/basic/path-util.c [hunk 9]
```diff
@@ -1421,10 +1416,9 @@ bool path_strv_contains(char **l, const char *path) {
...
+
```
Comment (poettering) [inline src/basic/path-util.c:1421]: mayb make j const

File: src/basic/path-util.c (hunk 6 above)
Comment (medhefgo) [reply src/basic/path-util.c:206]: I was implying that we can reuse s instead of declaring t.

**Comment by poettering** [comment]
i cherry-picked the middle PRs into git main, they make sense on their own.

PR looks pretty good to me, see my minor points above. Great work!

File: src/basic/path-util.c (hunk 6 above)
Comment (yuwata) [reply src/basic/path-util.c:206]: Actually, here `t` can be dropped and the iterator can be reused. But I prefer the current form for readability.

File: src/basic/list.h (hunk 5 above)
Comment (yuwata) [reply src/basic/list.h:142]: The macro `LIST_FOREACH_WITH_NEXT()` is explicitly used in udev-rules.c.

File: src/basic/path-util.c (hunk 7 above)
Comment (yuwata) [reply src/basic/path-util.c]: Also, this is just for readability.

File: src/basic/path-util.c (hunk 9 above)
Comment (yuwata) [reply src/basic/path-util.c:1421]: Done.

File: src/basic/strv.h (hunk 8 above)
Comment (yuwata) [reply src/basic/strv.h:137]: This should be safe, but I updated this to make safer. PTAL.

**Comment by yuwata** [comment]
Thank you for all comments.
Rebased, and `STRV_FOREACH_BACKWARD()` is updated to address @medhefgo's concern.
PTAL.

File: src/basic/strv.h (hunk 8 above)
Comment (medhefgo) [reply src/basic/strv.h:137]: (This is still decrementing `i` and then checking, which would break for `i` pointing at the 0th string.)

**Comment by medhefgo** [comment]
Also, would be nice if you could move `STRV_FOREACH` into `string-fundamental.h` at the same time, then I won't have to move it myself. :D

File: src/basic/path-util.c (hunk 6 above)
Comment (keszybz) [reply src/basic/path-util.c:206]: Yeah, it's better to keep it.

File: src/basic/strv.h (hunk 8 above)
Comment (keszybz) [reply src/basic/strv.h:137]: I think this is a valid point, and we should make sure to not do UB. But since this is a "preexisting condition", let's leave it for later.

**Comment by keszybz** [comment]
Let's merge this before we get conflicts. Some further follow-ups are certainly possible, but the changes done here stand on their own.

File: src/basic/strv.h (hunk 8 above)
Comment (keszybz) [reply src/basic/strv.h:137]: https://github.com/systemd/systemd/pull/22797
